### ISSUE TITLE
LoginButton failing w/ code 308 with swift 2.3/iOS10

### ISSUE THREAD

**Issue by b-raines** [facebook/facebook-sdk-swift#51]
This works when running in iOS9.3 simulator, but not in iOS10.
Every time I attempt to login using the login button, I get an error message of:
`The operation couldn’t be completed. (com.facebook.sdk.login error 308.)`

Sample code below:

```
class SignInViewController: UIViewController, LoginButtonDelegate {
    var fbLoginButton: LoginButton!

    override func viewDidLoad() {
        super.viewDidLoad()
        
        fbLoginButton = LoginButton(readPermissions: [.Email, .PublicProfile])
        fbLoginButton.delegate = self
    }

    func loginButtonDidCompleteLogin(loginButton: LoginButton, result: LoginResult) {
        switch result {
        case .Failed(let error):
            // User log in failed with error
        case .Cancelled:
            // User cancelled login
        case .Success:
            // Log in user
        }
    }
    
    func loginButtonDidLogOut(loginButton: LoginButton) {
        // Do some logout stuff
    }
}
```

**Comment by esetnik** [comment]
I have the exact same issue as @b-raines

**Comment by danmunoz** [comment]
Same issue with Xcode 8 GM, iOS 10 GM

**Comment by esetnik** [comment]
I found that you have to enable keychain sharing for any keychain access to function properly. 

http://stackoverflow.com/questions/38456471/secitemadd-always-returns-error-34018-in-xcode-8-beta-gm-in-ios-10-simulator

http://stackoverflow.com/questions/38689631/how-to-use-facebook-ios-sdk-on-ios-10/38799196#38799196

And finally this might be the hint as to whether or not it's an Apple bug:
https://developer.apple.com/library/ios/documentation/Security/Reference/keychainservices/#//apple_ref/c/func/SecItemAdd

```When you use Xcode to create an application, Xcode adds an application-identifier entitlement to the application bundle. Keychain Services uses this entitlement to grant the application access to its own keychain items. You can also add a keychain-access-groups entitlement to the application and, in the entitlement property list file, specify an array of keychain access groups to which the application belongs. The property list file can have any name you like (for example, keychain-access-groups.plist). The Xcode build variable CODE_SIGN_ENTITLEMENTS should contain the SRCROOT relative path to the entitlement property list file. The property list file itself should be a dictionary with a top-level key called keychain-access-groups whose value is an array of strings. If you add such a property-list file to the application bundle, then the access group corresponding to the application-identifier entitlement is treated as the last element in the access groups array. If you do not include the kSecAttrAccessGroup key in the attributes dictionary when you call the SecItemAdd function to add an item to the keychain, the function uses the first access group in the array by default. If there is no kSecAttrAccessGroup key in the attributes dictionary and there is no keychain-access-groups entitlement in the application bundle, then the access group of a newly created item is the value of the application-identifier entitlement.```

**Comment by nlutsenko** [comment]
@esetnik, thank you so much for a writeup here.
Looks indeed like with iOS 10 having keychain turned on in the app capabilities (having an entry in entitlements file) is a requirement for using keychain.
We are storing access token and user profile, which are sensitive user data in the application's keychain, so unauthorized access (like with reading from plain file) is prevented.

If you experience any issues, even after enabling keychain capabilities, please open a new issue.

**Comment by esetnik** [comment]
@nlutsenko Lets update the readme to inform users that enabling keychain sharing is a required step for the SDK to function properly as of iOS 10. It's dangerous because we previously had the sdk working properly and as of iOS 10 it just broke without warning. I still think this is an Apple bug given their documentation states that the keychain should work even without the `keychain-access-groups` entitlement in the application bundle.

I think lots of high profile iOS apps got burned by this issue when iOS 10 came out. I got logged out of Slack, Twitter and Instagram as soon as I updated to iOS 10.

**Comment by nlutsenko** [comment]
@esetnik, sounds good. Would happily review and merge the pull request that can address this.

**Comment by adamdahan** [comment]
Enabling Keychain Sharing only patches this for Simulator but still does not work on devices running iOS 10+. This is not a solution.

**Comment by EvgenyKarkan** [comment]
> Enabling Keychain Sharing only patches this for Simulator but still does not work on devices running iOS 10+. This is not a solution.

@adamdahan did you find solution?

**Comment by dariuxmx** [comment]
Enabling Keychain sharing does not run on device (iOS 10+).

**Comment by brandons** [comment]
Enabling keychain sharing does not fix this issue for me. Any other advice?

@adamdahan @EvgenyKarkan @dariuxmx @esetnik

**Comment by iwming** [comment]
Any other advice?

**Comment by aldrich** [comment]
Have to also follow up on this. I'm getting sporadic 308 errors on simulator and device (iOS 12) for the latest version, with or without Keychain Sharing enabled.

**Comment by rspl-nilesh** [comment]
I am also getting same error in simulator and device with latest facebook SDK version 4.38, with or without keychain sharing enabled.

**Comment by bazoooki** [comment]
Same here. @aldrich @rspl-nilesh 

if you find any workaround let us know, please. we are trying to debug it as well.

**Comment by mp3il** [comment]
same here.

**Comment by andytriboletti** [comment]
I am also getting same error in simulator and device with latest facebook SDK version 4.38, with or without keychain sharing enabled. Any other advice highly appriciated.

**Comment by nitinbizsingh** [comment]
I am also seeing sporadic 308 errors on version 4.31.1 of FBSDKLoginKit

**Comment by rinat-enikeev** [comment]
quick and dirty fix described here: https://github.com/facebook/facebook-sdk-swift/issues/286#issuecomment-433700139

**Comment by rspl-nilesh** [comment]
@bazoooki , @andytriboletti , We can use the SDK version 3.37 instead of latest until issue is been fixed and that version works like charm, if Bitcode is disable for the project.

**Comment by zrx89757** [comment]
getting the same error when we update the SDK to newest version,with Xcode 10,iOS10.x;
FIX IT when we rollback to v4.33.0,May this works for you !
